### PR TITLE
Catalyst/PennyLane adjoint unification

### PR DESCRIPTION
**Pull request by dime10**
Restructures adjoint to inherit from PL adjoint for the single op case. The multi op case is not 1:1 compatible because PL only ever instantiates the Adjoint class on a single base operation. This means that it returns a _list_ of `Adjoint` instances while Catalyst intends to return a single `Adjoint` instance with a list of ops nested inside of it.

Additionally:
- Rename Adjoint class to HybridAdjoint
- Allow adjoint to be used outside of qjit/qnode similar to PL allowing operations to be instantiated anywhere
- Refactor adjoint nested wrappers into a callable
- Add eager (`lazy=False`) adjoint support for single ops
- Include the PL test suite for adjoint in Catalyst to test the better integrated behaviour

[sc-60922]
[sc-60923]

### CODE REVIEW

File: frontend/test/pytest/test_adjoint.py
Comment (dime10) [reply frontend/test/pytest/test_adjoint.py:579]: I think we've been supporting it for a while, PL still tests both though hence this fixture.

**Review by erick-xanadu — APPROVED** [review]
I'm going to approve, but if you do add a decomposition, can you ping me again for review? Thanks!

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:521]: Can we use the `Adjoint.decomposition` method now?

File: frontend/test/pytest/test_adjoint.py
Comment (maliasadi) [inline frontend/test/pytest/test_adjoint.py:1667]: A thorough testing of adjoint :heart:

**Review by maliasadi — COMMENTED** [review]
Thanks @dime10! Nice work :100:

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:515]: As far as I know we've had no support for observables so far

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:530]: This class has more meta-data than its base class, so the PyTree methods need to be overridden

File: frontend/test/pytest/test_adjoint.py
Comment (maliasadi) [inline frontend/test/pytest/test_adjoint.py:30]: I cannot test this class locally without the following change,
```suggestion
class TestCatalyst:
```

Comment (dime10) [reply frontend/test/pytest/test_adjoint.py:30]: Good catch, thanks!

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:399]: Of course, the user could supply `catalyst.adjoint` with an arbitrary input

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:521]: For single ops I believe so

Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:521]: I tested this locally and we now use `Adjoint.decomposition` with single ops :partying_face:

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:53]: nice hack! :)

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:530]:
```suggestion
        # going to PennyLane's adjoint skips unnecessarily re-queuing the base operation in each
```

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:531]: Thanks for the comment! :+1:

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:581]: :tada:

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:571]: I may be missed it in `test_adjoint.py`; where is this adjoint(obs) tested?

**Review by maliasadi — APPROVED** [review]
Nice work! :partying_face:

File: frontend/catalyst/api_extensions/__init__.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/__init__.py:38]: Just a reminder to add this PR to the changelog before merging ;)

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:571]: Good question, I don't know. I would hope it's part of the PL test suite that I copied over 👀

File: frontend/catalyst/api_extensions/__init__.py
Comment (dime10) [reply frontend/catalyst/api_extensions/__init__.py:38]: Good point!

File: frontend/catalyst/device/decomposition.py
Comment (dime10) [reply frontend/catalyst/device/decomposition.py:40]: `api_extensions` as a submodule provides its own public methods and objects to be used by the rest of the catalyst package. In such a scheme modules outside of the `api_extensions` folder (not counting tests) should not need to use objects not available directly from the `api_extesions`.

File: frontend/catalyst/api_extensions/__init__.py
Comment (dime10) [reply frontend/catalyst/api_extensions/__init__.py:38]: Done 👍

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Done

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:571]: It seems to be tested in a few places such as [here](https://github.com/PennyLaneAI/catalyst/blob/fc1b0320ed347fa9f69c68e506c7155f87db2899/frontend/test/pytest/test_adjoint.py#L1201). There are also tests for HermitianObs :+1:

File: frontend/catalyst/api_extensions/quantum_operators.py
Comment (maliasadi) [inline frontend/catalyst/api_extensions/quantum_operators.py:243]: minor:
```suggestion
    # Return an instantiated version of the Adjoint class if we receive an operator instance.
    return adj() if isinstance(f, Operator) else adj
```

**Pull request merged**
